### PR TITLE
Features/backend migrate

### PR DESCRIPTION
**Pull request by nnabse**

### CODE REVIEW

File: src/app/services/http.service.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import { GLOBAL_LINK } from 'src/links.constants';
+import { DeleteResponse, Todo, UpdateResponse } from '../models/Todo';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class HttpService {
+  constructor(private http: HttpClient) {}
+
+  get(link: string, params?: any): Observable<Todo[]> {
```
Comment (MkpDolganovKirill) [inline src/app/services/http.service.ts:13]: You can create new type for "params" like HttpParams | { [p: string]: string | number | boolean | readonly (string | number | boolean)[] } | undefined

File: src/app/services/todo.service.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+@Injectable({
+  providedIn: 'root',
+})
+export class TodoService implements OnInit {
+  public todoList: Todo[] = [];
+
+  constructor(private httpService: HttpService) {}
+
+  ngOnInit(): void {}
```
Comment (MkpDolganovKirill) [inline src/app/services/todo.service.ts:22]: remove ngOnInit() and implements for it if you don't use that

File: src/app/services/todo.service.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+  getAllTasks(): void {
+    this.httpService
+      .get(GET_ALL_TASKS_LINK)
+      .subscribe((resp) => ((this.todoList = resp), console.log(resp)));
```
Comment (MkpDolganovKirill) [inline src/app/services/todo.service.ts:27]: remove console.log

File: src/app/services/http.service.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    return this.http.get<Todo[]>(`${GLOBAL_LINK}${link}`, { params });
+  }
+
+  create(link: string, body: object, params?: any): Observable<Todo> {
```
Comment (MkpDolganovKirill) [inline src/app/services/http.service.ts:17]: Use "generics" if you want to do this http requests universal

File: src/app/components/todo/todo.component.html [hunk 5]
```diff
@@ -12,12 +12,12 @@ <h3>Todo App</h3>
 </div>
 <div class="task_list">
   <div
-    *ngFor="let todo of TodoList"
+    *ngFor="let todo of todoService.todoList"
```
Comment (ViktorFenev) [inline src/app/components/todo/todo.component.html:15]: Do not use instance of service in HTML

File: src/app/models/Todo.ts [hunk 6]
```diff
@@ -1,6 +1,19 @@
...
+  deletedCount: number;
+}
+
+export interface UpdateResponse {
+  acknowledged: boolean;
+  modifiedCount: number;
+  upsertedId?: null;
+  upsertedCount: number;
+  matchedCount: number;
+}
```
Comment (ViktorFenev) [inline src/app/models/Todo.ts:19]: Is it still necessary?

File: src/app/services/todo.service.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    this.httpService
+      .delete(DELETE_TASK_LINK, { _id })
+      .subscribe(() => this.getAllTasks());
+  }
+
+  updateTask(_id: number, body: any): void {
+    this.httpService
+      .update(UPDATE_TASK_LINK, { _id }, body)
+      .subscribe(() => this.getAllTasks());
+  }
```
Comment (ViktorFenev) [inline src/app/services/todo.service.ts:46]: How will you catch errors?

File: src/app/services/todo.service.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  getAllTasks(): void {
+    this.httpService
+      .get(GET_ALL_TASKS_LINK)
+      .subscribe((resp) => ((this.todoList = resp), console.log(resp)));
+  }
+
+  createTask(body: object): void {
+    this.httpService
+      .create(CREATE_TASK_LINK, body)
+      .subscribe(() => this.getAllTasks());
```
Comment (ViktorFenev) [inline src/app/services/todo.service.ts:33]: Why you get all tasks after create only 1 task???

File: src/app/components/todo/todo.component.ts [hunk 9]
```diff
@@ -1,79 +1,72 @@
...
-  public renameTodoForm = {
-    newName: '',
-  };
+  constructor(public todoService: TodoService) {}
 
-  constructor() {}
+  ngOnInit(): void {
+    this.todoService.todoList.subscribe((data) => {
+      this.localTodoList = data;
+    });
```
Comment (MkpDolganovKirill) [inline src/app/components/todo/todo.component.ts:24]: You don't unsubscribe from BehaviorSubject in onDestroy

**Review by MkpDolganovKirill — APPROVED** [review]

File: src/app/services/http.service.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+})
+export class HttpService {
+  constructor(private http: HttpClient) {}
+
+  get<T>(link: string, params?: Params): Observable<T> {
+    return this.http.get<T>(`${GLOBAL_LINK}${link}`, { params });
+  }
+
+  create<T>(link: string, body: object, params?: Params): Observable<T> {
+    return this.http.post<T>(`${GLOBAL_LINK}${link}`, body, { params });
```
Comment (ViktorFenev) [inline src/app/services/http.service.ts:18]: Тип Params непонятен. Параметры чего и какие?

File: src/app/services/todo.service.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+  updateTask(_id: number, body: any): void {
+    this.httpService
+      .update(UPDATE_TASK_LINK, { _id }, body)
+      .pipe(catchError((err) => throwError(() => err)))
+      .subscribe(() => {
+        this.todoList.next(
+          this.todoList.value.map((elem) => {
+            if (_id === elem._id) {
+              return (elem = { ...elem, ...body });
```
Comment (ViktorFenev) [inline src/app/services/todo.service.ts:57]: return { ...elem, ...body } не делает то же самое?

File: src/app/components/todo/todo.component.ts [hunk 12]
```diff
@@ -1,79 +1,72 @@
...
+  private isEditorOpened: boolean = false;
 
-  public renameTodoForm = {
-    newName: '',
-  };
+  constructor(public todoService: TodoService) {}
 
-  constructor() {}
+  ngOnInit(): void {
+    this.todoService.todoList.subscribe((data) => {
```
Comment (ViktorFenev) [inline src/app/components/todo/todo.component.ts:22]: Subjects should be names with $ in the end of name like todoList$

And u should unsubscribe from this subject

File: src/app/components/todo/todo.component.ts [hunk 13]
```diff
@@ -1,79 +1,72 @@
...
+  private isEditorOpened: boolean = false;
 
-  public renameTodoForm = {
-    newName: '',
-  };
+  constructor(public todoService: TodoService) {}
 
-  constructor() {}
+  ngOnInit(): void {
+    this.todoService.todoList$.subscribe((data) => {
```
Comment (ViktorFenev) [inline src/app/components/todo/todo.component.ts:22]: I don't see unsubscribe

**Review by ViktorFenev — APPROVED** [review]

**Pull request merged**